project/design-patterns/src/prototype/ShapeCache.java

```
@@ -1,20 +1,22 @@
 package prototype;
 
-import java.util.Hashtable;
+import java.util.HashMap;
+import java.util.Map;
 
 public class ShapeCache {
-	private static Hashtable<String, Shape> shapeMap = new Hashtable<String, Shape>();
+	private static Map<String, Shape> shapeMap = new HashMap<String, Shape>();
 	
-	public static Shape getShape(String shapeId) {
-		Shape cachedShape = shapeMap.get(shapeId);
-		
-		return (Shape) cachedShape.clone();
+	static Shape getShape(String id) {
+		Shape cachedShape = shapeMap.get(id);
+		if(shapeMap.containsKey(cachedShape.getId())) {
+			return (Shape) cachedShape.clone();
+		}
+		return null;
```

Instead of returning null, throw your exception here.